### PR TITLE
feat: implement logger

### LINKED ISSUE #79: Add a `logger` module to integrate with Cloud Logging
Create a Python equivalent of the Node [`logger` subpackage](https://firebase.google.com/docs/functions/writing-and-viewing-logs) to integrate with Cloud Logging's structured logs

Also include trace; https://github.com/firebase/firebase-functions/blob/aae1abd55930550636e8e95c41feefcb617edc9d/src/common/trace.ts#L6

### PR DESCRIPTION
**Pull request by exaby73**
Closes #79

### CODE REVIEW

**Comment by taeold** [comment]
q: why does the example you shared in https://github.com/firebase/firebase-functions-python/pull/121#issuecomment-1734255387 emit a textPayload not a jsonPayload as I'd expect when you give the log a structured data like a dictionary?

**Comment by exaby73** [comment]
@taeold The code was adapted from the [node logger](https://github.com/firebase/firebase-functions/blob/3e7a4b77967e46a067712445c499c5df005b8e31/src/logger/index.ts)

**Comment by taeold** [comment]
@exaby73 I'm not sure I understand what you mean. The purpose of nodejs logger is so that it's easy to emit structured log entry from valid JS object and not to emit them as just text message.

e.g.

```

```

If the python logger isn't doing any work to generate structued logs using `jsonPayload` and can set appropriate logging levels that corresponds to logging levels in Cloud Logging, then it's not more useful that regular print() functions.

**Comment by exaby73** [comment]
@taeold I've updated the logger to support jsonPayload. The `*args` will be joined with a space to produce the `message` key and the `**kwargs` will be spread into the `jsonPayload`. Here is the result of calling `logger.log("Hello World!", foo="Bar")`:

image

File: src/firebase_functions/private/_identity_fn.py [hunk 1]
```diff
@@ -114,7 +114,7 @@ def _auth_user_record_from_token_data(token_data: dict[str, _typing.Any]):
     return AuthUserRecord(
         uid=token_data["uid"],
         email=token_data.get("email"),
-        email_verified=token_data.get("email_verified"),
+        email_verified=bool(token_data.get("email_verified")),
```
Comment (taeold) [inline src/firebase_functions/private/_identity_fn.py:117]: hmm this feels unrelated? should we make a separate PR for this or is this necessary for log impl?

File: src/firebase_functions/private/util.py [hunk 2]
```diff
@@ -387,6 +387,8 @@ def timestamp_conversion(time: str) -> _dt.datetime:
     elif precision_timestamp == PrecisionTimestamp.SECONDS:
         return second_timestamp_conversion(time)
 
+    raise ValueError("Invalid timestamp")
```
Comment (taeold) [inline src/firebase_functions/private/util.py:390]: likewise - related to log impl?

File: src/firebase_functions/private/_identity_fn.py (hunk 1 above)
Comment (exaby73) [reply src/firebase_functions/private/_identity_fn.py:117]: mypy was updated and was causing CI to fail because the types didn't match. I'm open to suggestions. This is mainly a linting fix

File: src/firebase_functions/private/util.py (hunk 2 above)
Comment (exaby73) [reply src/firebase_functions/private/util.py:390]: Again a lint failure here because the function wasn't ending in a return and the if, elif is not exhaustive

File: src/firebase_functions/private/_identity_fn.py (hunk 1 above)
Comment (antont) [reply src/firebase_functions/private/_identity_fn.py:117]: maybe clearest to make the mypy related linting fixes in a separate PR, and have only the logging things here.

Comment (exaby73) [reply src/firebase_functions/private/_identity_fn.py:117]: Without these, the CI doesn't pass. Possibly can make a separate PR for these that can merged into `main`, before we merge this in so that we can have those changes rebased into this branch. I'll look into doing this next week :)

Comment (taeold) [reply src/firebase_functions/private/_identity_fn.py:117]: eh let's save ourselves some time and push it in this PR.

**Review by taeold — APPROVED** [review]

**Comment by taeold** [comment]
@exaby73 thank you!!
